### PR TITLE
Add more information to `impl Trait` error

### LINKED ISSUE #92458: using `impl Trait` on a return type as part of a trait declaration results in a misleading error
I tried this code:

```rust
pub trait Test {
    fn into_iter(&self) -> impl Iterator<Item = Self> {
        todo!()  // A proper implementation will also fail
    }
}
```

I expected to see this happen: an error message stating that `impl Trait` is not allowed as part of trait declarations.

Instead, this happened: the following misleading error
```
error[E0562]: `impl Trait` not allowed outside of function and method return types
 --> src/main.rs:3:28
  |
3 |     fn into_iter(&self) -> impl Iterator<Item = Self> {
  |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0562`.
```
This error is misleading because it states that the problem is that the keyword `impl` is not being used as part of a return type, even though it is.

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (efec54529 2021-12-04)
binary: rustc
commit-hash: efec545293b9263be9edfb283a7aa66350b3acbf
commit-date: 2021-12-04
host: x86_64-pc-windows-msvc
release: 1.59.0-nightly
LLVM version: 13.0.0
```

<details><summary>Backtrace</summary>
<p>
Produces the exact same error

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #92458

Let me know if I went overboard here, or if the suggestions could use some refinement.

r? @estebank 
Feel free to reassign to someone else

### CODE REVIEW

**Comment by compiler-errors** [comment]
Alternatively we can disambiguate #92458 by just adding that word "inherent", without adding these `ImplTraitPosition`s everywhere..

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -1338,14 +1395,15 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
                         let mut err = struct_span_err!(
                             self.sess,
                             t.span,
                             E0562,
-                            "`impl Trait` not allowed outside of {}",
-                            "function and method return types",
+                            "`impl Trait` not allowed outside of \
+                            function and inherent method return types",
                         );
+                        err.note(&format!("found `impl Trait` in {}", position));
```
Comment (estebank) [inline compiler/rustc_ast_lowering/src/lib.rs:1406]: I'm concerned about "inherent" in this wording not being too clear. "associated functions" doesn't communicate the same meaning, sadly. Either way, I would likely change this to be 

```suggestion
                            "`impl Trait` only allowed in function and inherent method return types, not in {}", position
                        );
```

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -282,6 +291,54 @@ impl<'a> ImplTraitContext<'_, 'a> {
     }
 }
 
+impl std::fmt::Display for ImplTraitPosition {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let name = match self {
+            ImplTraitPosition::Path => "path",
+            ImplTraitPosition::Variable => "variable",
```
Comment (estebank) [inline compiler/rustc_ast_lowering/src/lib.rs:298]:
```suggestion
            ImplTraitPosition::Variable => "variable binding",
```

**Review by estebank — APPROVED** [review]
Couple of nitpicks, r=me after rebasing

**Comment by compiler-errors** [comment]
Rebased and addressed comments, ready for r+

**Pull request merged**